File: app/src/test/java/com/lambdasoup/watchlater/viewmodel/AddViewModelTest.java
/*
 *   Copyright (c) 2015 - 2017
 *
 *   Maximilian Hille <mh@lambdasoup.com>
 *   Juliane Lehmann <jl@lambdasoup.com>
 *
 *   This file is part of Watch Later.
 *
 *   Watch Later is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   Watch Later is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with Watch Later.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.lambdasoup.watchlater.viewmodel;

import android.accounts.Account;
import android.content.Intent;
import android.net.Uri;

import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.lambdasoup.watchlater.WatchLaterApplication;
import com.lambdasoup.watchlater.data.AccountRepository;
import com.lambdasoup.watchlater.data.YoutubeRepository;
import com.lambdasoup.watchlater.util.VideoIdParser;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestRule;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class AddViewModelTest {

	@Rule
	public TestRule rule = new InstantTaskExecutorRule();

	@Mock
	private WatchLaterApplication application;

	@Mock
	private AccountRepository accountRepository;

	@Mock
	private VideoIdParser videoIdParser;

	@Mock
	private YoutubeRepository youtubeRepository;

	private AddViewModel viewModel;

	private MutableLiveData<Account> accountLiveData;

	@Before
	public void setup() {
		when(application.getAccountRepository()).thenReturn(accountRepository);
		when(application.getVideoIdParser()).thenReturn(videoIdParser);
		when(application.getYoutubeRepository()).thenReturn(youtubeRepository);

		accountLiveData = new MutableLiveData<>();
		when(accountRepository.get()).thenReturn(accountLiveData);

		viewModel = new AddViewModel(application);
	}

	@Test
	public void should_set_account() {
		Account account = mock(Account.class);

		viewModel.setAccount(account);

		verify(accountRepository).put(account);
	}

	@SuppressWarnings("unchecked")
	@Test
	public void should_get_account() {
		LiveData<Account> liveData = mock(LiveData.class);
		when(accountRepository.get()).thenReturn(liveData);

		AddViewModel viewModel = new AddViewModel(application);
		LiveData<Account> defaultAccount = viewModel.getAccount();

		assertThat(defaultAccount).isEqualTo(liveData);
	}

	@Test
	public void should_remove_account() {
		viewModel.removeAccount();

		verify(accountRepository).clear();
	}

	@Test
	public void should_set_uri() {
		Uri uri = mock(Uri.class);
		when(videoIdParser.parseVideoId(uri)).thenReturn("test");

		viewModel.setVideoUri(uri);
	}

	@Test
	public void should_set_permission_needed() {
		viewModel.setPermissionNeeded(true);
		assertThat(viewModel.getPermissionNeeded().getValue()).isEqualTo(true);
	}

	@Test
	public void should_set_add_progress_when_permission_to_true() {
		viewModel.setPermissionNeeded(true);

		viewModel.setPermissionNeeded(false);

		AddViewModel.VideoAdd videoAdd = viewModel.getVideoAdd().getValue();
		assertThat(videoAdd).isNotNull();
		assertThat(videoAdd.state).isEqualTo(AddViewModel.VideoAdd.State.IDLE);
	}

	@Test
	public void should_error_when_watchlater_without_account() {
		viewModel.watchLater();

		AddViewModel.VideoAdd videoAdd = viewModel.getVideoAdd().getValue();
		assertThat(videoAdd).isNotNull();
		assertThat(videoAdd.state).isEqualTo(AddViewModel.VideoAdd.State.ERROR);
		assertThat(videoAdd.errorType).isEqualTo(AddViewModel.VideoAdd.ErrorType.NO_ACCOUNT);
	}

	@Test
	public void should_error_when_watchlater_without_permissions() {
		accountLiveData.setValue(mock(Account.class));
		viewModel.setPermissionNeeded(true);

		viewModel.watchLater();

		AddViewModel.VideoAdd videoAdd = viewModel.getVideoAdd().getValue();
		assertThat(videoAdd).isNotNull();
		assertThat(videoAdd.state).isEqualTo(AddViewModel.VideoAdd.State.ERROR);
		assertThat(videoAdd.errorType).isEqualTo(AddViewModel.VideoAdd.ErrorType.NO_PERMISSION);
	}

	@Test
	public void should_request_token_when_watchlater() {
		accountLiveData.setValue(mock(Account.class));
		viewModel.setPermissionNeeded(false);

		viewModel.watchLater();

		AddViewModel.VideoAdd videoAdd = viewModel.getVideoAdd().getValue();
		assertThat(videoAdd).isNotNull();
		assertThat(videoAdd.state).isEqualTo(AddViewModel.VideoAdd.State.PROGRESS);
		verify(accountRepository).getToken(viewModel);
	}

	@Test
	public void should_error_when_token_result_has_error() {
		viewModel.onToken(true, null, null);

		AddViewModel.VideoAdd videoAdd = viewModel.getVideoAdd().getValue();
		assertThat(videoAdd).isNotNull();
		assertThat(videoAdd.state).isEqualTo(AddViewModel.VideoAdd.State.ERROR);
		assertThat(videoAdd.errorType).isEqualTo(AddViewModel.VideoAdd.ErrorType.NO_ACCOUNT);
	}

	@Test
	public void should_intent_when_token_result_has_intent() {
		Intent intent = mock(Intent.class);
		viewModel.onToken(true, null, intent);

		AddViewModel.VideoAdd videoAdd = viewModel.getVideoAdd().getValue();
		assertThat(videoAdd).isNotNull();
		assertThat(videoAdd.state).isEqualTo(AddViewModel.VideoAdd.State.INTENT);
		assertThat(videoAdd.intent).isEqualTo(intent);
	}

	@Test(expected = IllegalStateException.class)
	public void should_throw_when_token_but_no_videoid() {
		viewModel.onToken(false, "test", null);
	}

	@Test
	public void should_add_video_when_token() {
		Uri uri = mock(Uri.class);
		when(videoIdParser.parseVideoId(uri)).thenReturn("video-id");
		viewModel.setVideoUri(uri);

		viewModel.onToken(false, "token", null);

		verify(youtubeRepository).addVideo("video-id", "token", viewModel);
	}

	@Test
	public void should_set_success() {
		viewModel.onAddResult(null, null);

		AddViewModel.VideoAdd videoAdd = viewModel.getVideoAdd().getValue();
		assertThat(videoAdd).isNotNull();
		assertThat(videoAdd.state).isEqualTo(AddViewModel.VideoAdd.State.SUCCESS);
	}

	@Test
	public void should_handle_invalid_token() {
		// first try -> clear token, retry
		viewModel.onAddResult(YoutubeRepository.ErrorType.INVALID_TOKEN, "token");
		verify(accountRepository).invalidateToken("token");
		verify(accountRepository).getToken(viewModel);

		// 2nd try -> error
		viewModel.onAddResult(YoutubeRepository.ErrorType.INVALID_TOKEN, "token2");
		AddViewModel.VideoAdd videoAdd = viewModel.getVideoAdd().getValue();
		assertThat(videoAdd).isNotNull();
		assertThat(videoAdd.state).isEqualTo(AddViewModel.VideoAdd.State.ERROR);
		assertThat(videoAdd.errorType).isEqualTo(AddViewModel.VideoAdd.ErrorType.OTHER);
	}

	@Test
	public void should_handle_already_in_playlist() {
		viewModel.onAddResult(YoutubeRepository.ErrorType.ALREADY_IN_PLAYLIST, null);

		AddViewModel.VideoAdd videoAdd = viewModel.getVideoAdd().getValue();
		assertThat(videoAdd).isNotNull();
		assertThat(videoAdd.state).isEqualTo(AddViewModel.VideoAdd.State.ERROR);
		assertThat(videoAdd.errorType).isEqualTo(AddViewModel.VideoAdd.ErrorType.YOUTUBE_ALREADY_IN_PLAYLIST);
	}

	@Test
	public void should_handle_generic_error() {
		viewModel.onAddResult(YoutubeRepository.ErrorType.NETWORK, null);

		AddViewModel.VideoAdd videoAdd = viewModel.getVideoAdd().getValue();
		assertThat(videoAdd).isNotNull();
		assertThat(videoAdd.state).isEqualTo(AddViewModel.VideoAdd.State.ERROR);
		assertThat(videoAdd.errorType).isEqualTo(AddViewModel.VideoAdd.ErrorType.OTHER);
	}

	@Test
	public void should_videoinfo_success() {
		YoutubeRepository.Videos.Item item = mock(YoutubeRepository.Videos.Item.class);
		viewModel.onVideoInfoResult(null, item);

		AddViewModel.VideoInfo videoInfo = viewModel.getVideoInfo().getValue();
		assertThat(videoInfo).isNotNull();
		assertThat(videoInfo.state).isEqualTo(AddViewModel.VideoInfo.State.LOADED);
		assertThat(videoInfo.data).isEqualTo(item);
	}

	@Test
	public void should_videoinfo_error() {
		viewModel.onVideoInfoResult(YoutubeRepository.ErrorType.VIDEO_NOT_FOUND, null);

		AddViewModel.VideoInfo videoInfo = viewModel.getVideoInfo().getValue();
		assertThat(videoInfo).isNotNull();
		assertThat(videoInfo.state).isEqualTo(AddViewModel.VideoInfo.State.ERROR);
		assertThat(videoInfo.error).isEqualTo(YoutubeRepository.ErrorType.VIDEO_NOT_FOUND);
	}
}

File: app/src/test/java/com/lambdasoup/watchlater/viewmodel/AddViewModelTest.kt
/*
 * Copyright (c) 2015 - 2022
 *
 * Maximilian Hille <mh@lambdasoup.com>
 * Juliane Lehmann <jl@lambdasoup.com>
 *
 * This file is part of Watch Later.
 *
 * Watch Later is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Watch Later is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Watch Later.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.lambdasoup.watchlater.viewmodel

import android.accounts.Account
import android.content.Intent
import android.net.Uri
import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import androidx.lifecycle.MutableLiveData
import com.google.common.truth.Truth.assertThat
import com.lambdasoup.tea.testing.TeaTestEngineRule
import com.lambdasoup.watchlater.data.AccountRepository
import com.lambdasoup.watchlater.data.YoutubeRepository
import com.lambdasoup.watchlater.data.YoutubeRepository.*
import com.lambdasoup.watchlater.data.YoutubeRepository.Playlists.Playlist
import com.lambdasoup.watchlater.util.VideoIdParser
import com.lambdasoup.watchlater.viewmodel.AddViewModel.*
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.whenever
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TestRule
import org.junit.runner.RunWith
import org.mockito.MockitoAnnotations.initMocks
import org.mockito.junit.MockitoJUnitRunner

@RunWith(MockitoJUnitRunner::class)
class AddViewModelTest {

    @get:Rule
    var aacRule: TestRule = InstantTaskExecutorRule()

    @get:Rule
    var teaRule: TestRule = TeaTestEngineRule()

    private val accountRepository: AccountRepository = mock()
    private val videoIdParser: VideoIdParser = mock()
    private val youtubeRepository: YoutubeRepository = mock()

    private lateinit var vm: AddViewModel
    private lateinit var accountLiveData: MutableLiveData<Account>
    private val playlistLiveData = MutableLiveData<Playlist?>()

    private val token = "token"
    private val uri: Uri = mock()
    private val videoId = "video-id"
    private val item: Videos.Item = mock()
    private val observedEvents = mutableListOf<Event>()
    private val playlist: Playlist = mock()

    @Before
    fun setup() {
        initMocks(this)

        accountLiveData = MutableLiveData()
        whenever(accountRepository.get()).thenReturn(accountLiveData)
        whenever(youtubeRepository.targetPlaylist).thenReturn(playlistLiveData)
        playlistLiveData.value = playlist

        vm = AddViewModel(accountRepository, youtubeRepository, videoIdParser)

        whenever(accountRepository.getAuthToken())
                .thenReturn(AccountRepository.AuthTokenResult.AuthToken(token))
        accountLiveData.value = mock()
        whenever(videoIdParser.parseVideoId(uri)).thenReturn(videoId)
        whenever(youtubeRepository.getVideoInfo(videoId, token))
                .thenReturn(VideoInfoResult.VideoInfo(item))
        whenever(youtubeRepository.addVideo(videoId, playlist, token))
                .thenReturn(AddVideoResult.Success)

        observedEvents.clear()

        vm.setVideoUri(uri)
        vm.setPermissionNeeded(false)
    }

    @Test
    fun `should set account`() {
        val account: Account = mock()

        vm.setAccount(account)

        verify(accountRepository).put(account)
    }

    @Test
    fun `should get account`() {
        val account: Account = mock()
        accountLiveData.value = account

        assertThat(vm.model.value!!.account).isEqualTo(account)
    }

    @Test
    fun `should set permission needed`() {
        vm.setPermissionNeeded(true)

        assertThat(vm.model.value!!.permissionNeeded).isEqualTo(true)
    }

    @Test
    fun `should set add progress when permission to true`() {
        vm.setPermissionNeeded(true)
        vm.setPermissionNeeded(false)

        val videoAdd = vm.model.value?.videoAdd
        assertThat(videoAdd).isNotNull()
        assertThat(videoAdd).isEqualTo(VideoAdd.Idle)
    }

    @Test
    fun `should error when watchlater without account`() {
        accountLiveData.value = null

        vm.watchLater(videoId)

        val videoAdd = vm.model.value!!.videoAdd
        assertThat(videoAdd).isNotNull()
        assertThat(videoAdd).isInstanceOf(VideoAdd.Error::class.java)
        assertThat((videoAdd as VideoAdd.Error).error).isEqualTo(VideoAdd.ErrorType.NoAccount)
    }

    @Test
    fun `should error when watchlater without permissions`() {
        accountLiveData.value = mock()

        vm.setPermissionNeeded(true)
        vm.watchLater(videoId)

        val videoAdd = vm.model.value!!.videoAdd
        assertThat(videoAdd).isNotNull()
        assertThat(videoAdd).isInstanceOf(VideoAdd.Error::class.java)
        assertThat((videoAdd as VideoAdd.Error).error).isEqualTo(VideoAdd.ErrorType.NoPermission)
    }

    @Test
    fun `should add to watchlater`() {
        vm.watchLater(videoId)

        val videoAdd = vm.model.value?.videoAdd
        assertThat(videoAdd).isEqualTo(VideoAdd.Success)
    }

    @Test
    fun `should error when token result has error`() {
        whenever(accountRepository.getAuthToken())
            .thenReturn(AccountRepository.AuthTokenResult.Error(AccountRepository.ErrorType.AccountRemoved))
        accountLiveData.value = mock()

        vm.watchLater(videoId)

        val videoAdd = vm.model.value!!.videoAdd
        assertThat(videoAdd).isNotNull()
        assertThat(videoAdd).isInstanceOf(VideoAdd.Error::class.java)
        assertThat((videoAdd as VideoAdd.Error).error).isEqualTo(VideoAdd.ErrorType.NoAccount)
    }

    @Test
    fun `should network error when token result has network error`() {
        whenever(accountRepository.getAuthToken())
            .thenReturn(AccountRepository.AuthTokenResult.Error(AccountRepository.ErrorType.Network))
        accountLiveData.value = mock()

        vm.watchLater(videoId)

        val videoAdd = vm.model.value!!.videoAdd
        assertThat(videoAdd).isNotNull()
        assertThat(videoAdd).isInstanceOf(VideoAdd.Error::class.java)
        assertThat((videoAdd as VideoAdd.Error).error).isEqualTo(VideoAdd.ErrorType.Network)
    }

    @Test
    fun `should intent when token result has intent`() {
        val intent: Intent = mock()
        whenever(accountRepository.getAuthToken())
                .thenReturn(AccountRepository.AuthTokenResult.HasIntent(intent))
        accountLiveData.value = mock()

        vm.watchLater(videoId)

        val videoAdd = vm.model.value!!.videoAdd
        assertThat(videoAdd).isNotNull()
        assertThat(videoAdd).isInstanceOf(VideoAdd.HasIntent::class.java)
        assertThat((videoAdd as VideoAdd.HasIntent).intent).isEqualTo(intent)

        assertTrue(vm.events.contains(Event.OpenAuthIntent(intent)))
    }

    @Test
    fun `should refresh token transparently`() {
        val token2 = "token2"

        whenever(youtubeRepository.addVideo(videoId, playlist, token))
            .thenReturn(AddVideoResult.Error(ErrorType.InvalidToken, token))
        whenever(youtubeRepository.addVideo(videoId, playlist, token2))
            .thenReturn(AddVideoResult.Success)

        whenever(accountRepository.getAuthToken())
                .thenReturn(AccountRepository.AuthTokenResult.AuthToken(token))
                .thenReturn(AccountRepository.AuthTokenResult.AuthToken(token2))

        vm.watchLater(videoId)

        verify(accountRepository).invalidateToken(token)
        val videoAdd = vm.model.value!!.videoAdd
        assertThat(videoAdd).isNotNull()
        assertThat(videoAdd).isEqualTo(VideoAdd.Success)
    }

    @Test
    fun `should error when token refresh fails`() {
        val token2 = "token2"

        whenever(youtubeRepository.addVideo(videoId, playlist, token))
            .thenReturn(AddVideoResult.Error(ErrorType.InvalidToken, token))
        whenever(youtubeRepository.addVideo(videoId, playlist, token2))
            .thenReturn(AddVideoResult.Error(ErrorType.InvalidToken, token2))

        whenever(accountRepository.getAuthToken())
                .thenReturn(AccountRepository.AuthTokenResult.AuthToken(token))
                .thenReturn(AccountRepository.AuthTokenResult.AuthToken(token2))

        vm.watchLater(videoId)

        verify(accountRepository).invalidateToken(token)
        val videoAdd = vm.model.value!!.videoAdd
        assertThat(videoAdd).isNotNull()
        assertThat(videoAdd).isEqualTo(VideoAdd.Error(VideoAdd.ErrorType.Other("InvalidToken")))
    }

    @Test
    fun `should handle generic error`() {
        whenever(youtubeRepository.addVideo(videoId, playlist, token))
            .thenReturn(AddVideoResult.Error(ErrorType.Other, token))

        vm.watchLater(videoId)

        val videoAdd = vm.model.value!!.videoAdd
        assertThat((videoAdd as VideoAdd.Error).error).isEqualTo(VideoAdd.ErrorType.Other("Other"))
    }

    @Test
    fun `should set videoid and get videoinfo`() {
        val uri: Uri = mock()
        val item: Videos.Item = mock()
        val videoId = "video-id"
        whenever(videoIdParser.parseVideoId(uri)).thenReturn(videoId)
        whenever(youtubeRepository.getVideoInfo(videoId, token))
            .thenReturn(VideoInfoResult.VideoInfo(item))

        vm.setVideoUri(uri)

        assertThat(vm.model.value!!.videoId).isEqualTo(videoId)
        assertThat(vm.model.value!!.videoInfo).isEqualTo(VideoInfo.Loaded(item))
    }

    @Test
    fun `should set videoinfo error`() {
        val uri: Uri = mock()
        val videoId = "video-id"
        whenever(videoIdParser.parseVideoId(uri)).thenReturn(videoId)
        whenever(youtubeRepository.getVideoInfo(videoId, token))
            .thenReturn(VideoInfoResult.Error(ErrorType.VideoNotFound))

        vm.setVideoUri(uri)

        assertThat(vm.model.value!!.videoId).isEqualTo(videoId)
        assertThat(vm.model.value!!.videoInfo)
            .isEqualTo(VideoInfo.Error(VideoInfo.ErrorType.Youtube(ErrorType.VideoNotFound)))
    }

    @Test
    fun `should reload videoinfo after error when new account is added`() {
        accountLiveData.value = null
        whenever(accountRepository.getAuthToken())
            .thenReturn(AccountRepository.AuthTokenResult.Error(AccountRepository.ErrorType.Network))

        vm.setVideoUri(uri)

        whenever(accountRepository.getAuthToken())
            .thenReturn(AccountRepository.AuthTokenResult.AuthToken(token))
        accountLiveData.value = mock()

        assertThat(vm.model.value!!.videoInfo)
            .isInstanceOf(VideoInfo.Loaded::class.java)
    }

    @Test
    fun `should remove videoAdd error when playlist gets changed`() {
        vm.selectPlaylist(mock())

        assertThat(vm.model.value!!.videoAdd)
            .isInstanceOf(VideoAdd.Idle::class.java)
    }
}
